src/routes/(blog-article)/profiling-the-torrust-bittorrent-udp-tracker/+page.md

```diff
@@ -0,0 +1,328 @@
+- The total amount of time spent running a function (`Incl.` column).
+- The total amount of time spent running a function excluding child functions (`Self` column).
+- Etcetera.
+
+But one of the most interesting things it's the interactive "Call Graph" where you can navigate through functions seeing how much time they spend.
+
+The main problem with this tool is the asynchronous code. Functions executed in tokio spawned tasks are not shown directly. Since the Torrust Tracker make an intense use of async code that makes it harder to reach conclusions from the tool.
+
+There are alternatives that we have not tried yet:
+
+1. Use other tools like [tokio-console](https://github.com/tokio-rs/console).
+2. You can also insert you custom profiling functions in the code.
+3. You can isolate the sync parts and profile them independently. For example, you could profile only the Torrent Repository or the Tracker domain Service.
+
+## Profiling With Perf
+
+Another alternative for profiling are flamegraphs. There is a crate called [cargo flamegraph](https://github.com/flamegraph-rs/flamegraph) that allows you to easily generate flamegraphs that you can visualize in a modern browser.
+
+Flamegraphs offer a visual representation of code execution, providing invaluable insights into performance bottlenecks within the tracker codebase. By utilizing flamegraph profiling, developers can pinpoint hotspots and inefficiencies, guiding optimization efforts for enhanced performance.
+
+After installing the `cargo flamegraph` package you can generate the flamegraph with:
+
+<CodeBlock lang="terminal">
+
+```terminal
+TORRUST_TRACKER_PATH_CONFIG="./share/default/config/tracker.udp.benchmarking.toml" cargo flamegraph --bin=profiling -- 60
+```
+
+</CodeBlock>
+
+Running the `cargo flamegraph` running the whole service produces a flamegraph like [this](/images/posts/profiling-the-torrust-bittorrent-udp-tracker/flamegraph.svg):
+
+![Visualizing profiling data with flamegraphs](/images/posts/profiling-the-torrust-bittorrent-udp-tracker/flamegraph.svg)
+
+[Click here to see the interactive flamegraph](/images/posts/profiling-the-torrust-bittorrent-udp-tracker/flamegraph.svg).
+
+As you can see, there are a lot functions for the tokio package but if you see at the top of the flame you will see the functions in the `torrust_tracker` namespace.
+
+## Conclusion
+
+Although using these tools with async code can be very challenging at least you can use them to confirm if you are making an effort on improving the performance on the right place. And you can also detect parts very slow parts.
+
+And as you can see more than 90% of the time is spent on the `Tracker::announce` method in these two functions:
+
+- `Tracker::update_torrent_with_peer_and_get_stats`.
+- `Tracker::get_torrent_peers_for_peer`.
+
+Which are the functions that need a lock to write and read in the `Torrent Repository`. And the code were we are focused on improving right now.
```

Comment:
Starting a sentence with "And" is not a good practice.
Also the "were" thereafter should say "where".